daemon/run.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	successfulEntries := 0
+	sampledErrs := sampled_err.Errors{SampleCount: 3}
+
+	// Start a consumer goroutine that will read witnesses from a channel and pass them on to a backend collector.
+	witnessChannel := make(chan *Witness, WITNESS_BUFFER_SIZE)
+	doneChannel := make(chan bool)
+	go func() {
+		defer func() { doneChannel <- true }()
+
+		// Create collector for ingesting the witnesses.
+		collector := trace.NewBackendCollector(serviceID, learnSessionID, learnClient, kgxapi.Inbound, cmdArgs.Plugins)
+
+		// TODO: Offer option to remove third-party trackers?
+
+		// Create a new stream ID for the witnesses we are about to process.
+		streamID := uuid.New()
+
+		for seqNum := 0; true; seqNum++ {
+			// Get the next witness.
+			witness, more := <-witnessChannel
+			if !more {
+				return
+			}
+
+			// Pass the witness to the collector.
+			entrySuccess := apispec.ProcessHAREntry(collector, streamID, seqNum, *witness, &sampledErrs)
+			if entrySuccess {
+				successfulEntries += 1
+			}
+		}
+	}()
+
+	// Process the witnesses in the request body.
+	numWitnessesParsed := 0
+	numWitnessesDropped := 0
+	jsonDecoder := json.NewDecoder(request.Body)
+	for seqNum := 0; true; seqNum++ {
+		// Deserialize the next witness.
+		// XXX How to prevent client from giving us extremely large witnesses?
+		var witness Witness
+		if err := jsonDecoder.Decode(&witness); err != nil {
+			if err == io.EOF {
+				// We've reached the end of the stream, which isn't really an error.
+				err = nil
+			}
+			break
+		}
+
+		numWitnessesParsed++
+
+		// Attempt to enqueue the witness.
+		select {
+		case witnessChannel <- &witness:
+		default:
+			numWitnessesDropped++
+		}
+	}
+
+	// Signal the end of the witness stream and wait for the witnesses to finish uploading.
+	close(witnessChannel)
+	<-doneChannel
+
+	// Log any errors that we encountered while processing the HAR entries.
+	if sampledErrs.TotalCount > 0 {
+		log.Printf("Encountered errors with %d HAR entries.\n", numWitnessesParsed-numWitnessesDropped-successfulEntries)
+		log.Printf("Sample errors:\n")
+		for _, e := range sampledErrs.Samples {
+			log.Printf("\t- %s\n", e)
+		}
+	}
+
+	type witnessDetails struct {
+		// How many were parsed as valid witnesses.
+		Parsed int `json:"parsed"`
+
+		// How many were dropped because the queue was full.
+		Drops int `json:"drops"`
+
+		// How many errors were encountered.
+		Errors int `json:"errors"`
+
+		// How many were processed successfully.
+		Successes int `json:"successes"`
+	}
+
+	type responseBody struct {
+		Message        string         `json:"message,omitempty"`
+		WitnessDetails witnessDetails `json:"witnessDetails,omitempty"`
+	}
+
+	witDetails := witnessDetails{
+		Parsed:    numWitnessesParsed,
+		Drops:     numWitnessesDropped,
+		Errors:    sampledErrs.TotalCount,
+		Successes: successfulEntries,
+	}
+
+	// If we encountered malformed witnesses or there were errors processing witnesses, return a "Bad Request" status.
```

Comment:
It would be good to distinguish between parsing errors (400 response) throttling errors (429 response) and internal errors (5xx response).  Throttling is particularly important, because the client can use that information to alter the sampling rate.

Reply:
Here are the differences between what you propose and what is implemented:
* Errors returned from `apispec.ProcessHAREntry` are lumped in with parsing errors and yield a 400 status.
* Throttling errors currently yield a 422 status.

I can make the 422→429 change. It wasn't clear to me that errors from `apispec.ProcessHAREntry` are internal errors. But if that's what they are, I can have them produce a 500 status.

Reply:
`apispec.ProcessHAREntry` can encounter different kinds of errors, although it's not set up right now to expose them in an actionable way.

The first kind are errors returned from the backend.  This is where throttling errors would come from, as well as (hopefully transient) 5xx errors.

The second are errors `apispec.ProcessHAREntry` itself generates, which are hopefully just validation errors.

Ideally, we would be able to distinguish between different kinds of errors and pass them back to client.  (Best practice in API design is to pass errors back, rather than e.g. doing retries yourself in the face of throttling.)

For now, using `422` as a catchall is fine.  We can add handling for throttling and other specific errors later, after the daemon and middleware are done.

Reply:
I'll leave this as is. We can improve this when `apispec.ProcessHAREntry` has finer grained error reporting.